clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1405,6 +1405,8 @@ Sema::DeviceDiagBuilder Sema::SYCLDiagIfDeviceCode(SourceLocation Loc,
          "Should only be called during SYCL compilation");
   FunctionDecl *CurFunction = dyn_cast<FunctionDecl>(getCurLexicalContext());
   DeviceDiagBuilder::Kind DiagKind = [this, CurFunction] {
+    if (!CurFunction)
```

Comment:
I was thinking... I wonder if this case needs some sort of lexical context associated with it.  For example, I think the curlexicalcontext can also be the containing type (in the case of a member variable initializer).  Perhaps this needs to be more generic.

Reply:
Mmm, hope I understand your comment correctly.
Maybe once we teach deferred diagnostics system to apply diagnostics not only to some functions, we will need to analyze which type of lexical context do we have now...
But for now it seems enough.

Reply:
Based on the title of the pull-request it seems that this should be within the scope of this pull request!

In reality, there are 3 different cases I can think of:

1- CurLexicalContext is TU/global/etc (maybe null?): Should diagnose immediately.
2- CurLexicalContext is a FunctionDecl: Should do what it does now, enqueuing a diagnostic if that function isn't currently being emitted, else diagnosing.
2- CurLexicalContext is a type (typically a record type): This likely means we're in a member initializer.  If we're a static member, diagnose immediately.  Else, diagnose only once the type is used.

Reply:
> Based on the title of the pull-request it seems that this should be within the scope of this pull request!

Hmm, not sure about it. It says "fix" not "redesign". Because `DeviceDiagBuilder` knows how to attach diagnostics only to a function, I assume it needs some redesign to make deferred diagnostics for other types of Lexical context.

> 1- CurLexicalContext is TU/global/etc (maybe null?): Should diagnose immediately.

Also not sure. We should diagnose something only if it will get into the device code. Otherwise it will prevent from writing valid host code. If some global uses something illegal, but it doesn't get into device code (this likely means that this global is not used inside the device code), it should be ok.

> CurLexicalContext is a type (typically a record type): This likely means we're in a member initializer. If we're a static member, diagnose immediately. Else, diagnose only once the type is used.
I also think if this type is not used in device code, it shouldn't produce immediate diagnostics. I experimented with it, and saw that if this type is not used in device code, static member also doesn't get into device code and shouldn't be a problem.

Reply:
> Hmm, not sure about it. It says "fix for SYCL" not "redesign common parts". 

It says to 'fix deferred diagnostics for SYCL'.  This is a case where they are broken for SYCL :) 

> Also not sure. We should diagnose something only if it will get into the device code. Otherwise it will prevent from writing valid host code. If some global uses something illegal, but it doesn't get into device code (this likely means that this global is not used inside the device code), it should be ok.

Global initializers need to run on program launch, so they should ALWAYS get into the device code.

>I also think if this type is not used in device code, it shouldn't produce immediate diagnostics. I experimented with it, and saw that if this type is not used in device code, static member also doesn't get into device code and shouldn't be a problem.

Interesting... Static initializers are supposed to run same as global initializers, immediately on launch.  That is likely a different bug!  That said, the deferred diagnostics likely need to accept RecordDecl as well as FunctionDecl (for member initializers), and emit when that type is emitted.

Reply:
> It says to 'fix deferred diagnostics for SYCL'. This is a case where they are broken for SYCL :)

But they also broken for any programming model which can use in-class member initializers for example.

> Interesting... Static initializers are supposed to run same as global initializers, immediately on launch. That is likely a different bug! That said, the deferred diagnostics likely need to accept RecordDecl as well as FunctionDecl (for member initializers), and emit when that type is emitted.

I can double check If I missed something (likely in new year).
But, I'd say it's SYCL specific behavior, not a bug. Otherwise the rule "use regular c++" in host code won't work...

Reply:
> But they also broken for any programming model which can use in-class member initializers for example.
And yet... Still broken for SYCL.  I'm mostly teasing that your patch topic promises a 'fix all the things (for SYCL)', yet you're attempting to put things out of the scope of that.  A valid response to this would be to limit the scope of the patch.

>But, I'd say it's SYCL specific behavior, not a bug. Otherwise the rule "use regular c++" in host code won't work...
Its a C++ language rule.  Static initializers are supposed to run before main.  The language treats them as not part of the type (except for how they are named), but as global variables that should be accessible anywhere.  MANY programing models depend on static initialization to 'just work' that way.

Reply:
> And yet... Still broken for SYCL. I'm mostly teasing that your patch topic promises a 'fix all the things (for SYCL)', yet you're attempting to put things out of the scope of that. A valid response to this would be to limit the scope of the patch.
> 

I just didn't mean that the patch fixes all the things, I'm sorry if I confused you by this. I tried to say that this fixes related code added for SYCL but maybe you are right and it doesn't look so.
I see that you made the patch with similar fixes, but I'd like to keep some changes from this patch to, so I suppose the new topic for this patch will depend on how we decide to merge these patches.

> Its a C++ language rule. Static initializers are supposed to run before main. The language treats them as not part of the type (except for how they are named), but as global variables that should be accessible anywhere. MANY programing models depend on static initialization to 'just work' that way.

But SYCL is not regular C++ for device :), I'm even not sure that it should be called C++ at all... There is no main, and only kernels and functions called by kernels are supposed to run on device.

Reply:
SYCL might be normal C++ on some special devices sharing a lot of features with the host processor, but this is clearly outside of plain SYCL 1.2.1.
Of course, for the host device, this is already regular C++.
Intel Xeon Phi seems like a good candidate for a device that could support normal C++ once you have fine-grain system SVM.
Otherwise, I have some old presentations at Khronos F2F about some compromise adding the concept of SYCL program storage and later the generalization with scopes.
While this is not normal C++, it provides some way to create some objects with some persistent storage relative to some scope.